File: plugins/harness/skills/harness-install/assets/uv/runtime/reporter.py
# -*- coding: utf-8 -*-

from __future__ import annotations

from collections.abc import Iterable
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from harness_check_rule import Finding, FindingEdit


def _is_safe_reporter_path(root: Path, file_path: str) -> bool:
    """
    Check whether a finding/edit file path is safe for reporter reads.

    Rejects empty, absolute, parent traversal, symlink components, paths
    whose real path escapes root, and non-regular files.

    :param root: Repository root path.
    :param file_path: Relative file path from a finding or edit.
    :returns: True when the path is safe to read.
    :rtype: bool
    """
    if not file_path:
        return False
    from pathlib import PurePosixPath

    parsed = PurePosixPath(file_path)
    if parsed.is_absolute():
        return False
    for part in parsed.parts:
        if part == ".." or part.startswith("-"):
            return False
    resolved = (root / file_path).resolve(strict=False)
    root_resolved = root.resolve()
    if resolved != root_resolved and not str(resolved).startswith(f"{root_resolved}/"):
        return False
    try:
        real = resolved.resolve()
        if real != root_resolved and not str(real).startswith(f"{root_resolved}/"):
            return False
        if not real.is_file():
            return False
    except (OSError, ValueError):
        return False
    probe = root
    for part in parsed.parts:
        if part == ".":
            continue
        probe = probe / part
        try:
            if probe.is_symlink():
                return False
        except (OSError, ValueError):
            break
    return True


def _safe_read_lines(root: Path, file_path: str) -> list[str] | None:
    """
    Safely read file lines for reporting, returning None on unsafe paths.

    :param root: Repository root path.
    :param file_path: Relative file path from a finding or edit.
    :returns: File lines when safe, None otherwise.
    :rtype: list[str] | None
    """
    if not _is_safe_reporter_path(root, file_path):
        return None
    resolved = (root / file_path).resolve(strict=False)
    try:
        if not resolved.is_file():
            return None
    except (OSError, ValueError):
        return None
    return resolved.read_text(encoding="utf-8").splitlines()


def severity_label(severity: str) -> str:
    """
    Map severity value to uppercase label for output.

    :param severity: Severity level string (ERROR/WARN/INFO).
    :rtype: str
    """
    return severity.upper()


def format_location_header(finding: "Finding") -> str:
    """
    Format the first line of a finding with location information.

    :param finding: The finding to format.
    :rtype: str
    """
    label = severity_label(finding.severity)
    msg_part = f"{finding.category}: {finding.message}"
    if finding.file is None:
        return f"[{label}] {msg_part}"
    if finding.start_line is None:
        return f"{finding.file} [{label}] {msg_part}"
    return f"{finding.file}:{finding.start_line}:{finding.start_column if finding.start_column is not None else 1} [{label}] {finding.category}: {finding.message}"


def format_code_snippet(root: Path, finding: "Finding") -> list[str]:
    """
    Format code snippet with context lines around offending line.

    :param root: Repository root path.
    :param finding: The finding with location info.
    :rtype: list[str]
    """
    if finding.file is None or finding.start_line is None:
        return []
    lines = _safe_read_lines(root, finding.file)
    if lines is None:
        return []
    line_idx = finding.start_line - 1
    if line_idx < 0 or line_idx >= len(lines):
        return []
    start_idx = max(0, line_idx - 1)
    end_idx = min(len(lines), line_idx + 2)
    result = []
    for offset, code_line in enumerate(lines[start_idx:end_idx]):
        actual_line_no = start_idx + offset + 1
        result.append(f"  {'>' if actual_line_no == finding.start_line else ' '} {str(actual_line_no).rjust(len(str(end_idx)))}  │ {code_line}")
    return result


def extract_removed_text(root: Path, edit: "FindingEdit") -> list[str]:
    """
    Extract the original text at the edit location from the source file.

    :param root: Repository root path.
    :param edit: Edit operation with location coordinates.
    :returns: Original text spanning the edit range, one entry per source line.
    :rtype: list[str]
    """
    file_lines = _safe_read_lines(root, edit.file)
    if file_lines is None:
        return []
    start_idx = edit.start_line - 1
    end_idx = edit.end_line - 1
    if start_idx < 0 or end_idx >= len(file_lines):
        return []
    removed: list[str] = []
    for i in range(start_idx, end_idx + 1):
        line = file_lines[i]
        if i == start_idx and i == end_idx:
            slice_text = line[
                min(edit.start_column - 1, len(line)) : min(edit.end_column, len(line))
            ]
        elif i == start_idx:
            slice_text = line[min(edit.start_column - 1, len(line)) :]
        elif i == end_idx:
            slice_text = line[: min(edit.end_column, len(line))]
        else:
            slice_text = line
        removed.append(slice_text)
    return removed


def format_fix_info(root: Path, finding: "Finding") -> list[str]:
    """
    Format fix information (Safety, Help, Before/After).

    :param root: Repository root path.
    :param finding: The finding with optional fix metadata.
    :rtype: list[str]
    """
    if finding.fix is None:
        return []
    result = [
        f"  Safety: {finding.fix.safety.value}",
        f"  Help: {finding.fix.description}",
    ]
    if finding.fix.edits:
        first_edit = finding.fix.edits[0]
        result.append("")
        result.append("  Before:")
        for line in extract_removed_text(root, first_edit):
            result.append(f"  - {line}")
        result.append("  After:")
        for line in (
            first_edit.replacement.split("\n") if first_edit.replacement else [""]
        ):
            result.append(f"  + {line}")
    return result


def render_finding(root: Path, finding: "Finding") -> list[str]:
    """
    Render a single finding in structured diagnostic format.

    :param root: Repository root path.
    :param finding: The finding to render.
    :rtype: list[str]
    """
    result = [format_location_header(finding)]
    snippet = format_code_snippet(root, finding)
    if snippet:
        result.append("")
        result.extend(snippet)
    fix_info = format_fix_info(root, finding)
    if fix_info:
        result.append("")
        result.extend(fix_info)
    return result


def render_findings(root: Path, findings: Iterable["Finding"]) -> list[str]:
    """
    Render all findings in structured diagnostic format.

    Returns list of output lines suitable for printing. Gracefully handles
    findings where location or fix data is absent.

    :param root: Repository root path.
    :param findings: Iterable of findings to render.
    :rtype: list[str]
    """
    findings_list = list(findings)
    if not findings_list:
        return ["OK"]
    output = []
    for i, finding in enumerate(findings_list):
        if i > 0:
            output.append("")
        output.extend(render_finding(root, finding))
    import harness_check_rule

    fixable_count = sum(
        1
        for f in findings_list
        if f.fix is not None and f.fix.safety == harness_check_rule.FixSafety.SAFE
    )
    summary = f"Checked {len(set(f.file for f in findings_list if f.file is not None))} file(s). {len(findings_list)} violation(s): {sum(1 for f in findings_list if f.severity == 'ERROR')} error, {sum(1 for f in findings_list if f.severity == 'WARN')} warn, {sum(1 for f in findings_list if f.severity == 'INFO')} info."
    if fixable_count > 0:
        summary += f" [*] {fixable_count} fixable."
    output.append("")
    output.append(summary)
    return output

File: plugins/harness/skills/harness-install/assets/bun/runtime/reporter.ts
#!/usr/bin/env bun
// -*- coding: utf-8 -*-
import { lstatSync, readFileSync, realpathSync, statSync } from "node:fs";
import { isAbsolute, normalize, resolve, sep } from "node:path";
import type { Finding } from "./rules/harness-check-rule";

interface SafePathResult {
    readonly resolved: string;
    readonly safe: boolean;
}

const hasLeadingDashComponent = (path: string): boolean => path.split(/[\\/]/).some((segment) => segment !== "" && segment.startsWith("-"));

const safeRoot = (root: string): string => realpathSync(resolve(root));

const resolveSafeManifestPath = (root: string, path: string): SafePathResult => {
    if (path === "" || isAbsolute(path)) {
        return { resolved: "", safe: false };
    }

    const normalized = normalize(path);
    if (
        normalized === "." ||
        normalized === ".." ||
        normalized.startsWith("..") ||
        normalized.startsWith(`..${sep}`)
    ) {
        return { resolved: "", safe: false };
    }

    if (hasLeadingDashComponent(normalized)) {
        return { resolved: "", safe: false };
    }

    const absolutePath = resolve(root, normalized);
    const resolvedRoot = safeRoot(root);
    if (absolutePath !== resolvedRoot && !absolutePath.startsWith(`${resolvedRoot}${sep}`)) {
        return { resolved: "", safe: false };
    }

    let probe = root;
    for (const segment of normalized.split(/[\\/]/)) {
        if (segment === "" || segment === ".") {
            continue;
        }
        probe = resolve(probe, segment);
        try {
            if (lstatSync(probe).isSymbolicLink()) {
                return { resolved: "", safe: false };
            }
        } catch {
            break;
        }
    }

    try {
        const real = realpathSync(absolutePath);
        if (real !== resolvedRoot && !real.startsWith(`${resolvedRoot}${sep}`)) {
            return { resolved: "", safe: false };
        }
        if (!statSync(real).isFile()) {
            return { resolved: "", safe: false };
        }
        return { resolved: real, safe: true };
    } catch {
        return { resolved: "", safe: false };
    }
};

const resolveSafeManifestFile = (root: string, path: string): string | null => {
    const resolved = resolveSafeManifestPath(root, path);
    return resolved.safe ? resolved.resolved : null;
};
/**
 * Renders a collection of validation findings in structured diagnostic format.
 * Handles both findings with location/fix metadata and locationless findings with only severity/category/message.
 *
 * @param root Absolute filesystem root for resolving relative file paths.
 * @param findings Read-only array of validation findings to render.
 * @returns Array of formatted output lines (without trailing newlines).
 */
export function renderFindings(root: string, findings: readonly Finding[]): readonly string[] {
    if (findings.length === 0) {
        return ["OK"];
    }
    const lines: string[] = [];
    const fileSet = new Set<string>();
    const severityCounts: Record<"ERROR" | "WARN" | "INFO", number> = {
        ERROR: 0,
        WARN: 0,
        INFO: 0,
    };
    const safeFixCount = findings.filter((f) => f.fix && f.fix.safety === "safe").length;
    findings.forEach((finding) => {
        const hasFile = finding.file !== undefined;
        const hasLocation = hasFile && finding.startLine !== undefined;
        severityCounts[finding.severity]++;
        if (hasFile) {
            fileSet.add(finding.file);
        }
        if (!hasFile) {
            lines.push(`[${finding.severity.toUpperCase()}] ${finding.category}: ${finding.message}`);
        } else if (!hasLocation) {
            lines.push(`${finding.file} [${finding.severity.toUpperCase()}] ${finding.category}: ${finding.message}`);
        } else {
            lines.push(
                `${finding.file}:${finding.startLine}:${finding.startColumn ?? 1} [${finding.severity.toUpperCase()}] ${finding.category}: ${finding.message}`,
            );
            const snippetLines = getSnippet(root, finding.file, finding.startLine);
            if (snippetLines.length > 0) {
                lines.push("");
                snippetLines.forEach((line) => {
                    lines.push(line);
                });
            }
        }
        if (finding.fix) {
            lines.push("");
            lines.push(`   Safety: ${finding.fix.safety}`);
            lines.push(`   Help: ${finding.fix.description}`);
            if (finding.fix.edits && finding.fix.edits.length > 0) {
                lines.push("");
                lines.push("   Before:");
                extractEditText(root, finding.fix.edits[0])
                    .split("\n")
                    .filter((line) => line)
                    .forEach((line) => {
                        lines.push(`   - ${line}`);
                    });
                lines.push("   After:");
                finding.fix.edits[0].replacement
                    .split("\n")
                    .filter((line) => line)
                    .forEach((line) => {
                        lines.push(`   + ${line}`);
                    });
            }
        }
        lines.push("");
    });
    lines.push(
        `Checked ${fileSet.size} file(s). ${findings.length} violation(s): ${severityCounts.ERROR} error, ${severityCounts.WARN} warn, ${severityCounts.INFO} info.${safeFixCount > 0 ? ` [*] ${safeFixCount} fixable` : ""}`,
    );
    return lines;
}

/**
 * Extracts code context lines surrounding the finding location.
 * Shows 1 line before and 1 line after the offending line, with proper formatting.
 *
 * @param root Absolute filesystem root for resolving relative paths.
 * @param file File path (absolute or relative to root).
 * @param lineNumber 1-indexed line number of the finding.
 * @returns Array of formatted context lines.
 */
function getSnippet(root: string, file: string, lineNumber: number): readonly string[] {
    const absolutePath = resolveSafeManifestFile(root, file);
    if (absolutePath === null) {
        return [];
    }
    const fileLines = readFileSync(absolutePath, "utf8").split("\n");
    if (lineNumber < 1 || lineNumber > fileLines.length) {
        return [];
    }
    const before = lineNumber > 1 ? lineNumber - 1 : null;
    const after = lineNumber < fileLines.length ? lineNumber + 1 : null;
    const maxLineNum = Math.max(before ?? 0, lineNumber, after ?? 0);
    const lineNumWidth = String(maxLineNum).length;
    return [
        ...(before !== null
            ? [`   ${String(before).padStart(lineNumWidth, " ")} │ ${fileLines[before - 1] ?? ""}`]
            : []),
        `  > ${String(lineNumber).padStart(lineNumWidth, " ")}  │ ${fileLines[lineNumber - 1] ?? ""}`,
        ...(after !== null ? [`   ${String(after).padStart(lineNumWidth, " ")} │ ${fileLines[after - 1] ?? ""}`] : []),
    ];
}

/**
 * Extracts the original text at an edit location from the source file.
 * Used to display the "Before:" section of a fix.
 *
 * @param root Absolute filesystem root for resolving relative paths.
 * @param edit The edit operation containing location and replacement text.
 * @returns The original text that would be replaced.
 */
function extractEditText(
    root: string,
    edit: { file: string; startLine: number; startColumn: number; endLine: number; endColumn: number },
): string {
    const absolutePath = resolveSafeManifestFile(root, edit.file);
    if (absolutePath === null) {
        return "";
    }
    const lines = readFileSync(absolutePath, "utf8").split("\n");
    if (edit.startLine < 1 || edit.startLine > lines.length) {
        return "";
    }
    return edit.startLine === edit.endLine
        ? lines[edit.startLine - 1].slice(edit.startColumn - 1, edit.endColumn)
        : [
              lines[edit.startLine - 1].slice(edit.startColumn - 1),
              ...Array.from({ length: edit.endLine - edit.startLine - 1 }, (_, idx) => lines[edit.startLine + idx]),
              lines[edit.endLine - 1].slice(0, edit.endColumn),
          ].join("\n");
}
